vms/proposervm/proposer/windower_test.go

```diff
@@ -201,3 +161,277 @@ func TestDelayChangeByChain(t *testing.T) {
+	}
+}
+
+func TestMinDelayForProposer(t *testing.T) {
+	require := require.New(t)
+
+	validatorIDs, vdrState := makeValidators(t, 10)
+	w := New(vdrState, subnetID, fixedChainID)
+
+	var (
+		dummyCtx            = context.Background()
+		chainHeight  uint64 = 1
+		pChainHeight uint64 = 0
+		slot         uint64 = 0
+	)
+
+	expectedDelays := map[ids.NodeID]time.Duration{
+		validatorIDs[0]:          1 * WindowDuration,
+		validatorIDs[1]:          15 * WindowDuration,
+		validatorIDs[2]:          0 * WindowDuration,
+		validatorIDs[3]:          5 * WindowDuration,
+		validatorIDs[4]:          10 * WindowDuration,
+		validatorIDs[5]:          18 * WindowDuration,
+		validatorIDs[6]:          12 * WindowDuration,
+		validatorIDs[7]:          3 * WindowDuration,
+		validatorIDs[8]:          23 * WindowDuration,
+		validatorIDs[9]:          2 * WindowDuration,
+		ids.GenerateTestNodeID(): MaxLookAheadWindow,
+	}
+
+	for nodeID, expectedDelay := range expectedDelays {
+		delay, err := w.MinDelayForProposer(dummyCtx, chainHeight, pChainHeight, nodeID, slot)
+		require.NoError(err)
+		require.Equal(expectedDelay, delay)
+	}
+}
+
+func TestTimeToSlot(t *testing.T) {
+	parentTime := time.Now()
+	tests := []struct {
+		timeOffset   time.Duration
+		expectedSlot uint64
+	}{
+		{
+			timeOffset:   -WindowDuration,
+			expectedSlot: 0,
+		},
+		{
+			timeOffset:   -time.Second,
+			expectedSlot: 0,
+		},
+		{
+			timeOffset:   0,
+			expectedSlot: 0,
+		},
+		{
+			timeOffset:   WindowDuration,
+			expectedSlot: 1,
+		},
+		{
+			timeOffset:   2 * WindowDuration,
+			expectedSlot: 2,
+		},
+	}
+	for _, test := range tests {
+		t.Run(test.timeOffset.String(), func(t *testing.T) {
+			slot := TimeToSlot(parentTime, parentTime.Add(test.timeOffset))
+			require.Equal(t, test.expectedSlot, slot)
+		})
+	}
+}
+
+// Ensure that the proposer distribution is within 3 standard deviations of the
+// expected value assuming a truly random binomial distribution.
+func TestProposerDistribution(t *testing.T) {
+	require := require.New(t)
+
+	validatorIDs, vdrState := makeValidators(t, 10)
+	w := New(vdrState, subnetID, fixedChainID)
+
+	var (
+		dummyCtx               = context.Background()
+		pChainHeight    uint64 = 0
+		numChainHeights uint64 = 100
+		numSlots        uint64 = 100
+	)
+
+	proposerFrequency := make(map[ids.NodeID]int)
+	for chainHeight := uint64(0); chainHeight < numChainHeights; chainHeight++ {
+		for slot := uint64(0); slot < numSlots; slot++ {
+			proposerID, err := w.ExpectedProposer(dummyCtx, chainHeight, pChainHeight, slot)
+			require.NoError(err)
+			proposerFrequency[proposerID]++
+		}
+	}
+
+	var (
+		totalNumberOfSamples      = numChainHeights * numSlots
+		probabilityOfBeingSampled = 1 / float64(len(validatorIDs))
+		expectedNumberOfSamples   = uint64(probabilityOfBeingSampled * float64(totalNumberOfSamples))
+		variance                  = float64(totalNumberOfSamples) * probabilityOfBeingSampled * (1 - probabilityOfBeingSampled)
```

Comment:
Recall the variance of a binomial {0, 1} is equal to `p * q`.

Reply:
I think the code is right but I know it as `np(1-p)`...assuming `q` is just notation I'm not familiar with.

Reply:
yeah `q` is just `1-p`